src/riak_kv_wm_timeseries_query.erl

```diff
@@ -126,371 +144,171 @@ forbidden(RD, Ctx) ->
+        {error, backend_timeout} ->
+            %% the eleveldb process did manage to send us a timeout
+            %% response
+            Resp = riak_kv_wm_ts_util:set_error_message(
+                     "backend_timeout", [], RD),
+            {false, Resp, Ctx};
+        {error, Reason} ->
+            Resp = riak_kv_wm_ts_util:set_error_message(
+                     "select query execution error: ~p", [Reason], RD),
+            {false, Resp, Ctx}
     end.
 
--spec process_post(#wm_reqdata{}, #ctx{}) -> ?CB_RV_SPEC.
-%% @doc Pass through requests to allow POST to function
-%%      as PUT for clients that do not support PUT.
-process_post(RD, Ctx) ->
-    accept_doc_body(RD, Ctx).
-
--spec accept_doc_body(#wm_reqdata{}, #ctx{}) -> ?CB_RV_SPEC.
-accept_doc_body(RD0, Ctx0) ->
-    case preexec(RD0, Ctx0) of
-        {true, RD, Ctx} ->
-            call_api_function(RD, Ctx);
-        FalseWithDetails ->
-            FalseWithDetails
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Helper functions
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+query_from_request(RD) ->
+    QueryStr = query_string_from_request(RD),
+    lager:log(info, self(), "query_from_request: ~p", [QueryStr]),
+    compile_query(QueryStr).
+
+query_string_from_request(RD) ->
+    case wrq:get_qs_value("query", RD) of
+        undefined ->
+            throw({query, "no query key in query string"});
+        Str ->
+            Str
     end.
 
--spec call_api_function(#wm_reqdata{}, #ctx{}) -> ?CB_RV_SPEC.
-call_api_function(RD, Ctx = #ctx{result = Result})
-  when Result /= undefined ->
-    lager:debug("Function already executed", []),
-    {true, RD, Ctx};
-call_api_function(RD, Ctx = #ctx{method = Method,
-                                 compiled_query = CompiledQry}) ->
-    case CompiledQry of
-        SQL = ?SQL_SELECT{} when Method == 'GET' ->
-            %% inject coverage context
-            process_query(SQL, RD, Ctx);
-        Other when (is_record(Other, ddl_v1)               andalso Method == 'POST') orelse
-                   (is_record(Other, riak_sql_describe_v1) andalso Method ==  'GET') ->
-            process_query(Other, RD, Ctx);
-        _Other ->
-            handle_error({inappropriate_sql_for_method, Method}, RD, Ctx)
+compile_query(QueryStr) ->
+    case riak_ql_parser:ql_parse(
+           riak_ql_lexer:get_tokens(QueryStr)) of
+        {error, Reason} ->
+            ErrorMsg = lists:flatten(io_lib:format("parse error: ~p", [Reason])),
+            throw({query, ErrorMsg});
+        {ddl, _ } = Res ->
+            Res;
+        {Type, Compiled} when Type==select; Type==describe ->
+            {ok, SQL} =  riak_kv_ts_util:build_sql_record(
+                           Type, Compiled, undefined),
+            {Type, SQL}
     end.
 
+%% @todo: should really be in riak_ql somewhere
+table_from_sql(#ddl_v1{table=Table})                    -> Table;
```

Comment:
I think we need to move this into riak_ql somewhere - it seems like something that should be contained inside riak_ql.